# Наивная модель
### векторизация слов на основе TfIdf и классификатор RandomForest

Классификация текста — классическая задача в области обработки естественного языка. В качестве наивной модел будем использовать подход на основе TF-IDF. 
До победы ембедингов данный подход был основным.
При таком подходе на качество классификации больше  влияют методы предобработки и особенности текста. Будем считать что в рамках diplom_EDA я сделал все необходимое в этом направлении.  

На входе у нас список из обращений в техническую поддержку и список размеченных классов. Далее делаем Pipeline, который включает в себя векторизацию слов на основе TfIdf и классификатор RandomForest.

## SETUP

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
# загружаем библиотеки
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.ensemble import RandomForestClassifier
from sklearn.pipeline import Pipeline

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline 



In [3]:
PATH = 'drive/MyDrive/data/diplom_final/'
# PATH = ''

## EDA ...

## Load preprocessing data

In [4]:
# загружаем предобработанные данные
df = pd.read_csv(PATH+'data_prepare.csv', sep=',')
df.head()

class  ... numbers
0  Поддержка товаров  ...      25
1               Fiji  ...      16
2    Export buildman  ...      12
3                API  ...      37
4     Личный Кабинет  ...       8

[5 rows x 3 columns]

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9811 entries, 0 to 9810
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   class        9811 non-null   object
 1   description  9811 non-null   object
 2   numbers      9811 non-null   int64 
dtypes: int64(1), object(2)
memory usage: 230.1+ KB


In [7]:
df['class'].nunique()

12

In [8]:
df.head()

class  ... numbers
0  Поддержка товаров  ...      25
1               Fiji  ...      16
2    Export buildman  ...      12
3                API  ...      37
4     Личный Кабинет  ...       8

[5 rows x 3 columns]

In [11]:
y = df['class']
X = df['description']

In [12]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.1, shuffle=True, random_state=42)

## Model

In [161]:
model_naive = Pipeline([
                     ('tfidf', TfidfVectorizer()),
                     ('clf', RandomForestClassifier())
                     ])

In [162]:
model_naive.fit(X_train, y_train)

Pipeline(memory=None,
         steps=[('tfidf',
                 TfidfVectorizer(analyzer='word', binary=False,
                                 decode_error='strict',
                                 dtype=<class 'numpy.float64'>,
                                 encoding='utf-8', input='content',
                                 lowercase=True, max_df=1.0, max_features=None,
                                 min_df=1, ngram_range=(1, 1), norm='l2',
                                 preprocessor=None, smooth_idf=True,
                                 stop_words=None, strip_accents=None,
                                 sublinear_tf=False,
                                 token_pattern='...
                 RandomForestClassifier(bootstrap=True, ccp_alpha=0.0,
                                        class_weight=None, criterion='gini',
                                        max_depth=None, max_features='auto',
                                        max_leaf_nodes=None, max_samples=None

## Evaluate

In [163]:
model_naive.predict(['прошу удалить отзыв о нашей компнии так он не соответвует действительности'])[0]

'Поддержка по отзывам'

In [164]:
print('Model Accuracy - train data:', model_naive.score(X_train, y_train))
print('Model Accuracy - test data:', model_naive.score(X_test, y_test))

Model Accuracy - train data: 1.0
Model Accuracy - test data: 0.8808553971486762


Результат "наивной" модели - 0.88.  
Возьмем это за основу.

Так как это "наивная" модель, то более тонкую настройку выполнять не будем.  
Но в качестве потенциальных направления можно рассматривать:  
* дальнейшая предобработка текста 
* подбор параметров TfIdf:
* попробовать другой классификатор (SGD, SVM , XGB итд.)